vernac/declare.ml

```diff
@@ -2448,11 +2454,12 @@ let solve_by_tac prg obls i tac =
 let solve_and_declare_by_tac prg obls i tac =
   match solve_by_tac prg obls i tac with
   | None -> None
-  | Some (t, ty, uctx) ->
+  | Some (body, types, uctx) ->
     let obl = obls.(i) in
     let poly = Internal.get_poly prg in
+    let uctx = UState.restrict uctx (universes_of_decl body types) in
```

Comment:
I think I would put this in solve_by_tac (after control_only_guard) not here

Reply:
The failure in `bug_15099.v` suggests that both should actually be put in `declare_obligation` so that they impact only the given obligation and not the final constant (due to the `set_uctx` at the end of `solve_and_declare_by_tac`).

I pushed a new version.

Reply:
doesn't seem to work?

Reply:
There were inconsistencies between the execution paths for auto/manual obligations, and admitted obligations. After unifying the paths, that is after ensuring that in all three cases the monomorphic universes are declared on the first obligation constant, then, the restriction seems to work.

There is more to do to unify the admitted and non-admitted cases, but that will be for another PR.